### PR TITLE
add events after ensure installed completes

### PR DESCRIPTION
**Pull request by mehalter**
This PR looks at adding some User `autocmd` events when the `ensure_installed` function is completed as well as success/fail events for users. The event names are pretty lengthy so I'm totally up to discussing what these should be named.

Let me know what you think! Thanks for all your great work on this plugin and package management ecosystem!

### CODE REVIEW

**Review by williamboman — COMMENTED** [review]
Cool! I'd like to understand the use-case/need first. Is this needed in order to sequence a set of operations of sorts?

File: lua/mason-lspconfig/ensure_installed.lua [hunk 1]
```diff
@@ -16,6 +19,8 @@ local function resolve_package(lspconfig_server_name)
 end
 
 return function()
```
Comment (williamboman) [inline lua/mason-lspconfig/ensure_installed.lua:21]: We could avoid a lot of the state-tracking code by running this in an async context. We'd wrap the entire function with `return a.scope(function () ...  end)` and then we can leverage the `a.wait_all` utility to block execution until all provided "promises" are resolved.

Here's a slimmed down example:

```lua
local a = require "mason-core.async"

---@param pkg Package
local function async_install(pkg)
    ---@async
    return function ()
        return a.wait(function (resolve)
            pkg:install({ version = version } ):once("closed", function ()
                resolve(pkg:is_installed())
            end)
        end)
    end
end

-- a.scope creates a new coroutine scope, allowing us to execute async functions in it
return a.scope(function ()
    local installations = {}
    for _, pkg in ipairs(packages) do
        table.insert(installations, async_install(pkg))
    end

    if not _.all_pass(_.identity, a.wait_all(installations)) then
        -- one or more packages failed
    else
        -- success very nice
    end
end)
```

Comment (williamboman) [reply lua/mason-lspconfig/ensure_installed.lua:21]: I think it'd be best to keep the amount of events low. Is it enough with simply one event when installation completes? We can pass data to the event containing which packages were installed - then it's up to the consumer to do with that information whatever they want (e.g. check for failures)

**Comment by mehalter** [comment]
@williamboman I'm the core maintainer of AstroNvim and I'm trying to improve the installation process. The main thing I'm running into right now is mainly providing a way to indicate to users that things have been fully installed and their editor can be restarted in the case of a fresh installation that automatically installs servers.

The second part is for automatic installation with servers being installed in the background, there's no way to see when they complete or if any failed.

One note that's separate from this, I did notice that you are leveraging the notifications from installation finishing from Mason. I think these notifications only appear when the Mason window is open, so for automatic LSP installation we get a notification of "X is being installed" but no notification ever comes saying "server is successfully installed" (or failed).

TL;DR the main use case I'm after for these events is on a fresh installation of AstroNvim where the user has plugins/mason packages automatically installing, I would like to provide a notification or message to the user along the lines of "everything has been installed and you can now restart"

File: lua/mason-lspconfig/ensure_installed.lua (hunk 1 above)
Comment (mehalter) [reply lua/mason-lspconfig/ensure_installed.lua:21]: Ah all of this is great! I also agree that having one event is ideal in this case and totally meets the user case. Thanks for the guidance. I'll update my PR on Monday!

**Comment by mehalter** [comment]
Another use care which would be nice if this type of event is in place would be the option for users to easily open Mason if an automatically installed package fails to install.

**Comment by williamboman** [comment]
I see! Hm maybe it'd be best if you wrote something entirely custom that is tailored to your needs? The `ensure_installed` feature really only exists in mason-lspconfig because it was carried over from nvim-lsp-installer in order to keep feature parity. It's a very simple feature that I don't intend on maintaining long-term (eventually something similar would exist in core). For these reasons I'm also a bit hesitant about adding these events specifically for `ensure_installed`.

There's a somewhat simple API reference doc [here](https://github.com/williamboman/mason.nvim/blob/main/doc/reference.md) that you may use. By writing something custom you'd also be able to target more packages than just LSP servers. Replicating `ensure_installed` would require just 10-15 lines of code.

**Comment by mehalter** [comment]
Yeah I've thought about this as well. Thanks for the consideration and all the great work!
